File: #2616.minimize-the-maximum-difference-of-pairs.cpp
class Solution {
public:
    int minimizeMax(vector<int>& nums, int p) {
        
        std::sort(nums.begin(), nums.end());
        int size = nums.size();
        if (size == 1) {
            return 0;
        }
        
        long long low = 0;
        long long high = 1e10;
        long long mid;
        while (low < high) {
            mid = (low + high) / 2;
            vector<int> dp(size);
            dp[0] = 0;
            dp[1] = (nums[1]-nums[0]) <= mid;
            
            for (int i=2; i < size; i++) {
                dp[i] = std::max(((nums[i] - nums[i-1]) <= mid)
                    + dp[i-2], dp[i-1]);
            }
            
            if (dp[size - 1] >= p) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
};

File: #2616.minimize-the-maximum-difference-of-pairs.js
/**
 * @param {number[]} nums
 * @param {number} p
 * @return {number}
 */
var minimizeMax = function(nums, p) {
     nums.sort((a, b) => { return (a - b) });
     let n = nums?.length;
     let start = 0,end = nums[n -1] - nums[0];
     while (start < end) {
         let mid = Math.floor((start + end) / 2);
         let k = 0;
         for (let i = 1; i < nums?.length && k < p; i++) {
             if ((nums[i] - nums[i-1]) <= mid) {
                 k++;
                 i++;
             }   
         }

         if (k >= p) {
             end = mid;
         } else{
             start = mid + 1;
         }
     } 
     return start;
};
